### PR TITLE
Export "zarr/core" package export without codecs

### PR DESCRIPTION
**Pull request by manzt**
I've been experimenting with some ideas to enable a dynamic codec registry. The idea here is to export a "batteries-included" version of `zarr`, which is one large bundle and easy to use, and a `zarr-core` module which includes an empty registry that can be customized.

Some feedback on how/what to export at the top level of each bundled export would be useful. In the end I'd like to see the most minimal change to enable custom behavior of loading codecs.

```
165k zarr-core.es6.js
829k zarr.es6.js
```

### CODE REVIEW

**Comment by manzt** [comment]
@kylebarron, if you are interested. This change would allow others to create a registry how ever they prefer when using `zarr/core`.

```javascript
import { addCodec, openArray } from 'zarr/core';

// from unpkg
addCodec('blosc', () => import('https://unpkg.com/numcodecs@0.1.1/dist/blosc.mjs'));

// from node_modules
import { Zlib } from 'numcodecs';
addCodec(Zlib.codecId, () => Zlib);

// workerized
addCodec('gzip', () => /* wrap codec in worker pool and expose encode/decode */);


z = await zarr.openArray({ source: "http://localhost:8000" });
```

**Comment by kylebarron** [comment]
I like this approach because the developer can still import `zarr` itself for ease of use, or import just the core so that the cost of blosc is only needed when/if such a zarr dataset is actually requested.

**Comment by manzt** [comment]
Couple of updates here. Bundlers don't understand Node's package exports (yet), but the bundler resolvers for `rollup` and `webpack` will only look for files relative to the top level of package.

I've made some changes to our build configuration:

1.) The files from `dist` are published at the root of the npm package when running `npm publish`. This is kind of a hack, and there are several options [here](https://stackoverflow.com/questions/38935176/how-to-npm-publish-specific-folder-but-as-package-root). This means that how node resolves the module and how the bundlers resolve modules will be the same:

```javascript
// index.mjs -- Node
import { addCodec, openArray } from 'zarr/core'; // Node uses package exports

// index.js -- Webpack/Rollup
import { addCodec, openArray } from 'zarr/core'; // bundlers just try to match `./node_modlues/zarr/core*
```

2.) Only the ES module is code-split to expose `zarr/core` submodule. The commonjs and umd builds are still single file bundles. Since the resolvers for the bundlers don't check submodule exports, there are name conflicts with `core.cjs` and `core.mjs` that can't be resolved.

3.) I minified the `umd` bundle.

I think this strategy is the best way forward. We expose the same top-level exports (batteries included) `zarr.cjs` and `zarr.umd.js`, and the `zarr.mjs` is split as an ESM module.

**Comment by manzt** [comment]
Final bundles:

```bash
157k core.mjs (unminified)
665k zarr.mjs (unminified)
822k zarr.cjs (unminified)
539k zarr.umd.js (minified)
```

**Review by gzuidhof — APPROVED** [review]
Sorry for the long delay in reviewing this PR..

Thank you for splitting it up like this - it makes a lot of sense and the interface is simple even for the core version 👍. This PR is ready to merge when you are!

I think the core size is still pretty large (did we really write 100k characters? maybe..), I'll add a minified version for each of the `.mjs` releases in a later PR :)

File: package.json [hunk 1]
```diff
@@ -9,12 +9,33 @@
...
+    ".": {
+      "umd": "./zarr.umd.js",
+      "import": "./zarr.mjs",
+      "require": "./zarr.cjs"
+    },
+    "./core": {
+      "import": "./core.mjs"
+    }
+  },
   "author": "Guido Zuidhof <me@guido.io>",
```
[the remark this answers is not in the document]
Comment (manzt) [reply package.json:39]: I _think_ only one author is allowed for a package, but I added myself to the contributors!

**Comment by manzt** [comment]
Thansk guido! :)

File: docs/installation.md [hunk 2]
```diff
@@ -43,17 +43,17 @@ Importing as an ES module is the more modern and preferred approach. If you are
 
 ```js
 // Import invidual classes and functions.
-import { HTTPStore, openArray, slice } from "https://unpkg.com/zarr/dist/zarr.es6.js";
+import { HTTPStore, openArray, slice } from "https://cdn/skypack.dev/zarr";
```
Comment (kylebarron) [inline docs/installation.md:46]:
```suggestion
import { HTTPStore, openArray, slice } from "https://cdn.skypack.dev/zarr";
```

**Pull request merged**